### PR TITLE
Phase2: Digital river features: Invoice links, manage my cards, tax calculation

### PR DESCRIPTION
**Pull request by aavila-vinneren**
This PR contains new features:
1. Invoice links in order details page
2. Manage my credit cards from digital river under my account 
3. Tax calculation inclusive or exclusive

### CODE REVIEW

**Review by polishq — CHANGES_REQUESTED** [review]

File: react/Admin.tsx [hunk 1]
```diff
@@ -58,34 +64,8 @@ const Admin: FC = () => {
...
-    //     )
-    //     setSettingsState({
-    //       ...parsedSettings,
-    //       enableTaxCalculation: false,
-    //     })
-    //   }
-    // }
-  }, [data, account, formatMessage])
+    setSettingsState({ ...settingsState, ...parsedSettings })
+  }, [data, account, formatMessage, settingsState])
```
Comment (polishq) [inline react/Admin.tsx:68]: Having `settingsState` in the array here is causing an endless loop bug on the admin page. Basically this is saying that whenever `settingsState` changes, run the above code, but the above code changes `settingsState`. To fix, you can remove `settingsState` from the array on line 68 and change line 67 to this:

```
setSettingsState((prevState) => {
      return { ...prevState, ...parsedSettings }
    })
```

**Review by polishq — CHANGES_REQUESTED** [review]

File: node/middlewares/tax.ts [hunk 2]
```diff
@@ -1,283 +1,270 @@
...
-      price: item.itemPrice,
-      discount: item.discountPrice
-        ? {
-            amountOff: Math.abs(item.discountPrice / item.quantity),
-            quantity: item.quantity,
-          }
-        : undefined,
-      metadata: {
-        taxHubRequestId: item.id,
-      },
```
Comment (polishq) [inline node/middlewares/tax.ts:93]: Why were these lines removed? It's especially important to include any discounts on the item, or the tax will be calculated incorrectly. I would make sure the items array is being constructed the same as in `middlewares/checkout.ts` to ensure that calculations are consistent.

Additionally, we recently encountered a bug in another app having to do with free gifts. I tested the scenario with this new version of DR and tax was calculated for the free item as if it were being sold at its full price. To handle this, I suggest checking for the `item.isGift` flag and manually setting its price as 0 in the DR array.

**Review by polishq — CHANGES_REQUESTED** [review]

File: react/CreditCardsPage.tsx [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+      )
+    })
+  }
+
+  useEffect(() => {
+    if (!data?.appSettings?.message) return
+
+    const parsedSettings = JSON.parse(data.appSettings.message)
+
+    loadDigitalRiver(parsedSettings.digitalRiverPublicKey)
```
[the remark this answers is not in the document]
Comment (aavila-vinneren) [reply react/CreditCardsPage.tsx:165]: Done!

File: node/middlewares/tax.ts (hunk 2 above)
Comment (aavila-vinneren) [reply node/middlewares/tax.ts]: Done!

File: react/Admin.tsx (hunk 1 above)
Comment (aavila-vinneren) [reply react/Admin.tsx]: Done!

File: react/Admin.tsx [hunk 4]
```diff
@@ -298,34 +287,35 @@ const Admin: FC = () => {
...
                 />
-              </section> */}
+              </section>
               <section className="pt4">
                 <Button
                   variation="primary"
                   onClick={() => handleSaveSettings(showToast)}
                   isLoading={settingsLoading}
                   disabled={
+                    (appId !== 'vtexus.connector-digital-river' &&
```
[the remark this answers is not in the document]
Comment (aavila-vinneren) [reply react/Admin.tsx:318]: Done!

File: react/Admin.tsx [hunk 5]
```diff
@@ -208,6 +167,36 @@ const Admin: FC = () => {
             }
           >
             <PageBlock>
+              {appId !== 'vtexus.connector-digital-river' && appId !== '' && (
+                <section className="pb4">
+                  <Alert type="error">
+                    {`${formatMessage({
+                      id: 'admin/digital-river.settings.alert.errorText',
+                    })} ${appId}`}
```
[the remark this answers is not in the document]
Comment (aavila-vinneren) [reply react/Admin.tsx]: Done!

File: node/middlewares/tax.ts [hunk 6]
```diff
@@ -1,283 +1,270 @@
...
+          error: err,
         })
       }
     }
 
-    taxes.push({
-      id: item.metadata?.taxHubRequestId ?? index.toString(),
-      taxes: detailsTaxes,
+    taxesResponse.push({
+      id: '0',
```
[the remark this answers is not in the document]
Comment (aavila-vinneren) [reply node/middlewares/tax.ts]: Done!

File: node/middlewares/tax.ts [hunk 7]
```diff
@@ -1,283 +1,270 @@
...
+          error: err,
+          checkoutId: checkoutResponse.id,
+          message: 'DigitalRiverOrderTaxHandler-deleteCheckout',
+        })
 
-      if (item.fees.taxAmount > 0) {
-        detailsTaxes.push({
-          name: `FEE TAX`,
-          value: item.fees.taxAmount,
+        throw new ResolverError({
```
[the remark this answers is not in the document]
Comment (aavila-vinneren) [reply node/middlewares/tax.ts]: Done!

File: messages/en.json [hunk 8]
```diff
@@ -6,15 +6,18 @@
...
   "admin/digital-river.settings.isLive.label": "Enable production mode",
   "admin/digital-river.settings.isLive.helpText": "Disable to use test mode, enable for production mode; make sure to use the appropriate token for the selected mode.",
   "admin/digital-river.settings.isAutomaticSync.label": "Enable automatic catalog sync",
   "admin/digital-river.settings.isAutomaticSync.helpText": "Enable automatic syncing of VTEX catalog changes to Digital River's system.",
-  "admin/digital-river.settings.enableTaxCalculation.label": "Enable tax calculation",
-  "admin/digital-river.settings.enableTaxCalculation.helpText": "Activate Digital River in checkout as your tax calculator. WARNING: Make sure to test checkout after changing this setting.",
+  "admin/digital-river.settings.alert.errorText": "To set digital river as a your payment gateway uninstall the tax provider",
+  "admin/digital-river.settings.enableTaxInclusive.label": "Enable tax inclusive",
+  "admin/digital-river.settings.enableTaxInclusive.helpText": "Activate tax inclusive for tax calculations. WARNING: Make sure to test checkout after changing this setting.",
```
[the remark this answers is not in the document]
Comment (aavila-vinneren) [reply messages/en.json]: Done!

File: messages/en.json [hunk 9]
```diff
@@ -6,15 +6,18 @@
...
   "admin/digital-river.settings.isLive.label": "Enable production mode",
   "admin/digital-river.settings.isLive.helpText": "Disable to use test mode, enable for production mode; make sure to use the appropriate token for the selected mode.",
   "admin/digital-river.settings.isAutomaticSync.label": "Enable automatic catalog sync",
   "admin/digital-river.settings.isAutomaticSync.helpText": "Enable automatic syncing of VTEX catalog changes to Digital River's system.",
-  "admin/digital-river.settings.enableTaxCalculation.label": "Enable tax calculation",
-  "admin/digital-river.settings.enableTaxCalculation.helpText": "Activate Digital River in checkout as your tax calculator. WARNING: Make sure to test checkout after changing this setting.",
+  "admin/digital-river.settings.alert.errorText": "To set digital river as a your payment gateway uninstall the tax provider",
+  "admin/digital-river.settings.enableTaxInclusive.label": "Enable tax inclusive",
```
[the remark this answers is not in the document]
Comment (aavila-vinneren) [reply messages/en.json]: Done!

File: messages/en.json [hunk 10]
```diff
@@ -6,15 +6,18 @@
...
   "admin/digital-river.settings.vtexAppToken.label": "VTEX App Token",
   "admin/digital-river.settings.vtexAppToken.helpText": "Enter a VTEX app key and token pair for this account. If you are configuring a marketplace-seller relationship, use a key and token that have permissions for all related accounts.",
   "admin/digital-river.settings.isLive.label": "Enable production mode",
   "admin/digital-river.settings.isLive.helpText": "Disable to use test mode, enable for production mode; make sure to use the appropriate token for the selected mode.",
   "admin/digital-river.settings.isAutomaticSync.label": "Enable automatic catalog sync",
   "admin/digital-river.settings.isAutomaticSync.helpText": "Enable automatic syncing of VTEX catalog changes to Digital River's system.",
-  "admin/digital-river.settings.enableTaxCalculation.label": "Enable tax calculation",
-  "admin/digital-river.settings.enableTaxCalculation.helpText": "Activate Digital River in checkout as your tax calculator. WARNING: Make sure to test checkout after changing this setting.",
+  "admin/digital-river.settings.alert.errorText": "To set digital river as a your payment gateway uninstall the tax provider",
```
[the remark this answers is not in the document]
Comment (aavila-vinneren) [reply messages/en.json]: Done!

File: manifest.json [hunk 11]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "connector-digital-river",
   "vendor": "vtexus",
-  "version": "1.3.0",
+  "version": "1.3.1",
```
[the remark this answers is not in the document]
Comment (aavila-vinneren) [reply manifest.json]: Done!

File: react/InvoicesPage.tsx [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+  const getInvoices = () => {
+    return data.map((invoice: any) => {
+      return (
+        <div className="mb3">
+          <Card>
+            <div className="flex justify-between">
+              <div className="lh-copy">
+                <div className="mb3 fw5">{`#${invoice.orderId}`}</div>
+                <div>{`Order date: ${invoice.orderDate}`}</div>
```
Comment (polishq) [inline react/InvoicesPage.tsx:54]: Currently this date is displayed like "2021-08-24T15:58:42.4641087+00:00". Can it be formatted similar to how dates are shown in My Orders?

**Review by polishq — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 13]
```diff
@@ -7,6 +7,15 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+## [2.3.0] - 2021-11-18
```
Comment (polishq) [inline CHANGELOG.md:10]: Please remove this line, the version number and date will be automatically populated when the PR is merged. For now the changelog should look like: 

```
## [Unreleased]

### Added

- Change 1
- Change 2
- etc
```

File: docs/README.md [hunk 14]
```diff
@@ -20,18 +20,20 @@ This app integrates Digital River with VTEX checkout, allowing shoppers to inter
...
 2. In your admin sidebar, access the **Other** section and click on `Digital River` and then on `Configuration`.
-3. In the settings fields, enter your `Digital River token`, `VTEX App Key` and `VTEX App Token`. For initial testing, use a test `Digital River token` and leave the `Enable production mode` toggle turned off. Turn on the `Enable automatic catalog sync` toggle to enable syncing of SKUs from VTEX to Digital River each time a SKU is added or updated in your VTEX catalog.
+3. In the settings fields, enter your `Digital River public key`, `Digital River token`, `VTEX App Key` and `VTEX App Token`. For initial testing, use a test `Digital River token` and leave the `Enable production mode` toggle turned off. Turn on the `Enable automatic catalog sync` toggle to enable syncing of SKUs from VTEX to Digital River each time a SKU is added or updated in your VTEX catalog. Turn on the `Enable tax inclusive prices` toggle to enable if your catalog uses tax-inclusive product prices.
```
Comment (polishq) [inline docs/README.md:23]:
```suggestion
3. In the settings fields, enter your `Digital River public key`, `Digital River token`, `VTEX App Key` and `VTEX App Token`. For initial testing, use a test `Digital River token` and leave the `Enable production mode` toggle turned off. Turn on the `Enable automatic catalog sync` toggle to enable syncing of SKUs from VTEX to Digital River each time a SKU is added or updated in your VTEX catalog. Turn on the `Enable tax inclusive prices` toggle if your catalog uses tax-inclusive product prices.
```

File: docs/README.md [hunk 15]
```diff
@@ -20,18 +20,20 @@ This app integrates Digital River with VTEX checkout, allowing shoppers to inter
...
 
 ⚠️ _For multiple accounts configured in a marketplace-seller relationship, the same `VTEX App Key` and `VTEX App Token` should be used for all of the accounts in which the app is installed. You can use any of the accounts to generate the key/token, and then grant additional permissions to the key/token by [creating a new user](https://help.vtex.com/en/tutorial/managing-users--tutorials_512) on each of the other accounts using the `VTEX App Key` in place of the user's email address, and then assigning the Owner role to that user._
 
+⚠️ _There should be no other tax calculation provider enabled, if so it will re required to disable it._
```
Comment (polishq) [inline docs/README.md:27]:
```suggestion
⚠️ _Digital River provides its own tax calculation service. Therefore there should be no other tax calculation provider enabled on your account. If there is, you will be required to disable it before Digital River can be configured._
```

File: react/CreditCardsPage.tsx [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+    return sources?.map((source: any) => {
+      return (
+        <div className="mb3">
+          <label className="mr3" htmlFor={source.id}>{`${
+            source.creditCard.brand
+          } ending with ${
+            source.creditCard.lastFourDigits
+          } expires ${`0${source.creditCard.expirationMonth}`.slice(-2)}/${
+            source.creditCard.expirationYear
+          }`}</label>
```
Comment (polishq) [inline react/CreditCardsPage.tsx:151]: Please use intl here

**Review by polishq — CHANGES_REQUESTED** [review]

File: node/middlewares/tax.ts [hunk 17]
```diff
@@ -91,193 +40,215 @@ export async function digitalRiverOrderTaxHandler(
...
-
-  logger.info({
-    message: 'DigitalRiverTaxCalculation-TaxHubResponse',
-    taxHubResponse: { itemTaxResponse: taxes, hooks: [] },
-  })
+  }
 
   ctx.body = {
-    itemTaxResponse: taxes,
+    itemTaxResponse: taxesResponse,
```
Comment (polishq) [inline node/middlewares/tax.ts:251]: Please log this response body for debugging purposes

File: node/middlewares/tax.ts [hunk 18]
```diff
@@ -91,193 +40,215 @@ export async function digitalRiverOrderTaxHandler(
...
+        checkoutPayload,
+        message: 'DigitalRiverOrderTaxHandler-createCheckout',
       })
     }
 
-    if (item.fees.amount > 0) {
-      detailsTaxes.push({
-        name: `FEES`,
-        value: item.fees.amount,
+    if (checkoutResponse) {
```
Comment (polishq) [inline node/middlewares/tax.ts:169]: Please log the checkoutResponse here

**Review by polishq — CHANGES_REQUESTED** [review]

File: node/middlewares/tax.ts [hunk 19]
```diff
@@ -245,13 +261,19 @@ export async function digitalRiverOrderTaxHandler(
         })
       }
     }
+
+    setCache({ ctx, hash: hashRequest, value: taxesResponse, ttl: 'SHORT' })
   }
 
+  logger.info({
+    message: 'DigitalRiverOrderTaxHandler',
+    taxesResponse,
```
Comment (polishq) [inline node/middlewares/tax.ts:270]:
```suggestion
    cacheResponse || taxesResponse,
```

**Review by polishq — CHANGES_REQUESTED** [review]

**Review by polishq — DISMISSED** [review]

**Review by polishq — DISMISSED** [review]

**Review by polishq — APPROVED** [review]

**Pull request merged**